File: matlab/calc_volume.m
function volume = calc_volume(gridd, contour_points, params, output_file)
    % calculate the volume
    % input:
    %   gridd - struct with X, Y, Z_surface, Z_base
    %   params - struct with contour_level, grid_size
    %   contour_points - boundary points coordinates
    %   output_file - path to output file for grid volumes
    % output:
    %   volume - struct with exca_volume, fill_volume, grid_volume
    %
    %   openEarthwork: An Open Source Library for Excavation Calculation
    %   Author(s): Hubery H.B. Woo (hbw8456@163.com)
    %   Copyright 2009-2024 Chongqing Three Gorges University

    % extract the coordinates
    X = gridd.X;
    Y = gridd.Y;
    Z_surface = gridd.Z_surface;
    Z_base = gridd.Z_base;
    grid_size = params.grid_size;
    
    exca_volume = 0;
    fill_volume = 0;
    
    % create a new figure for displaying the volume calculation results
    subplot(2, 2, 4);
    title('volume calculation');
    hold on;
    
    % plot the boundary
    plot(contour_points(:,1), contour_points(:,2), 'g-', 'LineWidth', 2);
    
    % plot the grid points
    plot(X, Y, 'k.', 'MarkerSize', 1);
    
    % open output file for writing grid volumes
    fid = fopen(output_file, 'w');
    fprintf(fid, 'X,Y,Volume\n'); % write header
    
    for i = 1:size(X, 1)
        for j = 1:size(X, 2)
            % check if the grid center is inside the boundary (the smaller the grid, the more accurate)
            if inpolygon(X(i,j), Y(i,j), contour_points(:,1), contour_points(:,2))
                if ~isnan(Z_surface(i, j)) && ~isnan(Z_base(i, j))
                    delta_z = Z_surface(i, j) - Z_base(i, j);
                    grid_area = grid_size^2;
                    current_volume = delta_z * grid_area;
                    
                    % write grid volume to file
                    fprintf(fid, '%.3f,%.3f,%.3f\n', X(i,j), Y(i,j), current_volume);
                    
                    % calculate the total volume
                    if delta_z >= 0
                        exca_volume = exca_volume + current_volume;
                    else
                        fill_volume = fill_volume + abs(current_volume);
                    end
                    
                    % add the volume label
                    add_label(X(i,j), Y(i,j), current_volume);
                end
            end
        end
    end
    
    % close the output file
    fclose(fid);
    
    grid_volume = exca_volume - fill_volume;
    
    % store the results
    volume.exca_volume = exca_volume;
    volume.fill_volume = fill_volume;
    volume.grid_volume = grid_volume;

    % show the volume calculation results in the figure
    text(0.5, 0.5, sprintf('exca volume: %.2f\nfill volume: %.2f\ngrid volume: %.2f', ...
        exca_volume, fill_volume, grid_volume), ...
        'Units', 'normalized', 'VerticalAlignment', 'middle', 'HorizontalAlignment', 'center');
    
    % set the figure properties
    title('Volume Calculation Results');
    xlabel('X Coordinate');
    ylabel('Y Coordinate');
    axis equal;
    hold off;
    
    % add the legend
    legend('boundary', 'grid points', 'Location', 'best');

end % function calc_volume

function add_label(x, y, volume)
    % add the volume label
    % input:
    %   x, y - label position coordinates
    %   volume - volume value
    
    % set the display threshold, only display significant volume changes
    threshold = 10; % can be adjusted according to needs
    
    if abs(volume) > threshold
        % determine the text color and label
        if volume > 0
            color = [1 0 0]; % excavation is red
            label = sprintf('+%.0f', volume); % display positive values
        else
            color = [0 0 1]; % fill is blue
            label = sprintf('%.0f', volume); % display negative values
        end
        
        % add the text label
        text(x, y, label, ...
            'Color', color, ...
            'HorizontalAlignment', 'center', ...
            'VerticalAlignment', 'middle', ...
            'FontSize', 8, ...
            'FontWeight', 'bold');
    end

end % function add_label

File: fortran/calc_volume.f90
subroutine calc_volume(grid_x, grid_y, z_surface, z_base, nx, ny, &
    contour_x, contour_y, n_contour, grid_size, &
    exca_volume, fill_volume, grid_volume)
    ! Calculate the volume
    ! input:
    !   grid_x, grid_y   - grid coordinates
    !   z_surface, z_base - surface and base elevations
    !   nx, ny           - grid dimensions
    !   contour_x, contour_y - boundary points
    !   n_contour        - number of contour points
    !   grid_size        - size of grid cell
    ! output:
    !   exca_volume      - excavation volume
    !   fill_volume      - fill volume
    !   grid_volume      - net volume
    !
    !   openExcav: An Open Source Library for Excavation Calculation
    !   Author(s): Hubery H.B. Woo (hbw8456@163.com)
    !   Copyright 2009-2024 Chongqing Three Gorges University
    
    implicit none
    
    ! Arguments
    integer, intent(in) :: nx, ny, n_contour
    real, intent(in) :: grid_x(nx,ny), grid_y(nx,ny)
    real, intent(in) :: z_surface(nx,ny), z_base(nx,ny)
    real, intent(in) :: contour_x(n_contour), contour_y(n_contour)
    real, intent(in) :: grid_size
    real, intent(out) :: exca_volume, fill_volume, grid_volume
    
    ! Local variables
    integer :: i, j
    real :: delta_z, grid_area, current_volume
    logical :: is_inside
    
    ! Initialize volumes
    exca_volume = 0.0
    fill_volume = 0.0
    grid_area = grid_size * grid_size
    
    ! Calculate volumes for each grid cell
    do i = 1, nx
        do j = 1, ny
            ! Check if grid point is inside boundary
            call check_point_in_polygon(grid_x(i,j), grid_y(i,j), &
                contour_x, contour_y, n_contour, is_inside)
            
            if (is_inside) then
                if (.not. isnan(z_surface(i,j)) .and. &
                    .not. isnan(z_base(i,j))) then
                    
                    delta_z = z_surface(i,j) - z_base(i,j)
                    current_volume = delta_z * grid_area
                    
                    ! Add to appropriate volume
                    if (delta_z >= 0.0) then
                        exca_volume = exca_volume + current_volume
                    else
                        fill_volume = fill_volume + abs(current_volume)
                    end if
                end if
            end if
        end do
    end do
    
    ! Calculate net volume
    grid_volume = exca_volume - fill_volume

end subroutine calc_volume

! Helper function to check if point is NaN
logical function isnan(x)
    real, intent(in) :: x
    isnan = (x /= x)
end function isnan

! Helper function to check if point is inside polygon
subroutine check_point_in_polygon(x, y, poly_x, poly_y, n, inside)
    implicit none
    real, intent(in) :: x, y
    integer, intent(in) :: n
    real, intent(in) :: poly_x(n), poly_y(n)
    logical, intent(out) :: inside
    
    integer :: i, j
    inside = .false.
    j = n
    
    do i = 1, n
        if ((poly_y(i) > y) .neqv. (poly_y(j) > y)) then
            if (x < (poly_x(j) - poly_x(i)) * (y - poly_y(i)) / &
                (poly_y(j) - poly_y(i)) + poly_x(i)) then
                inside = .not. inside
            end if
        end if
        j = i
    end do
end subroutine check_point_in_polygon 